core/templates/services/translate.service.ts

```diff
@@ -0,0 +1,137 @@
+ */
+
+import { Injectable, EventEmitter } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
+import { UtilsService } from './utils.service';
+
+export interface MissingTranslationHandlerParams {
+   // the key that's missing in translation files
+  key: string;
+
+  // an instance of the service that was unable to translate the key.
+  translateService: TranslateService;
+
+   // Interpolation params that were passed along for translating the given key.
+  interpolateParams?: Object;
+}
+
+export interface LangChangeEvent {
+  lang: string;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TranslateService {
+  currentLang = 'en';
+  defaultLang = 'en';
+  prefix = '/assets/i18n/';
+  suffix = '.json';
+  translations: Array<Object> = [];
+  templateMatcher: RegExp = /\<\[\s?([^{}\s]*)\s?\]\>/g;
+
+  private _onLangChange = new EventEmitter<LangChangeEvent>();
+  get onLangChange(): EventEmitter<LangChangeEvent> {
+    return this._onLangChange;
+  }
+
+  constructor(private http: HttpClient, private utils: UtilsService) {}
+
+  // This function fecthes translation.
+  fetchTranslations(lang:string) {
+    return this.http.get(`${this.prefix}${lang}${this.suffix}`).toPromise();
+  }
+
+  // This function sets the new translations
+  use(lang: string) {
+    // Check if it has already been fetched before
+    this.currentLang = lang;
+    if (Object.keys(this.translations).includes(lang)) {
+      this._onLangChange.emit(
+        {lang: lang });
+      return;
+    }
+    // Otherwise fetch the translations
+    this.translations[lang] = this.fetchTranslations(lang).then(
+      translations => {
+        this.translations[lang] = translations;
+        if (this.currentLang === lang) {
+          this._onLangChange.emit(
+            {lang: lang});
+        }
+      }
+    );
+  }
+
+
+  interpolateI18nString(expr: string | Function, params?: Object): string {
+    let interpolatedString: string;
+
+    if (typeof expr === 'string') {
+      interpolatedString = this.interpolateString(expr, params);
+    } else if (typeof expr === 'function') {
+      interpolatedString = this.interpolateFunction(expr, params);
+    } else {
+      interpolatedString = expr as string;
+    }
+
+    return interpolatedString;
+  }
+
+  private interpolateFunction(fn: Function, params?: Object | undefined) {
+    return fn(params);
+  }
+
+  private interpolateString(expr: string, params?: Object | undefined) {
+    if (!params) {
+      return expr;
+    }
+
+    return expr.replace(this.templateMatcher,
+      (substring: string, b: string) => {
+        let r = params[b];
+        return this.utils.isDefined(r) ? r : substring;
+      });
+  }
+
+  getInterpolatedString(
+      key: string,
+      interpolateParams?: Object): string {
+    if (!this.utils.isDefined(key) || !key.length) {
+      throw new Error('Parameter "key" required');
+    }
+    // check if we are loading a new translation to use {
```

Comment:
I don't know what this means or what { represents.

Reply:
Removed this comment!